### PR TITLE
(PDB-2815) Look for global configuration file

### PR DESCRIPTION
**Pull request by ajroetker**
This commit adds the ability to use a global configuration file with the
PuppetDB CLI, which we will fall back to if values in the per-user
config file are missing. This global config file is located at
`/etc/puppetlabs/client-tools/puppetdb.conf` similar to the location for
the global config for the other pe-client-tools.

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -125,7 +153,7 @@ impl PdbConfigSection {
                 .unwrap_or(&json::Json::Null) {
                     &json::Json::Array(ref urls) => Some(urls.into_iter().map(|url| url.as_string().unwrap().to_string()).collect::<Vec<String>>()),
                     &json::Json::String(ref urls) => Some(split_server_urls(urls.clone())),
-                    &json::Json::Null => Some(default_server_urls()),
+                    &json::Json::Null => None,
```
Comment (ajroetker) [inline src/config.rs]: We already do this defaulting elsewhere and it just gets in the way here when trying to do fallback correctly (merging configs).

File: src/config.rs [hunk 2]
```diff
@@ -100,9 +125,9 @@ fn default_server_urls() -> Vec<String> {
     vec!["http://127.0.0.1:8080".to_string()]
 }
 
-fn default_pdb_config_section() -> PdbConfigSection {
+fn empty_pdb_config_section() -> PdbConfigSection {
     PdbConfigSection {
-        server_urls: Some(default_server_urls()),
+        server_urls: None,
```
Comment (ajroetker) [inline src/config.rs]: We already do this defaulting elsewhere and it just gets in the way here when trying to do fallback correctly (merging configs).

**Comment by ajroetker** [comment]
So I think another code-path we could use for this is a `merge_configs` function which merges `global, user, flags` all together @senior not sure what looks better though.

File: README.md [hunk 3]
```diff
@@ -97,7 +97,9 @@ The Rust PuppetDB CLI accepts a `--config=<path_to_config>` flag which allows
 you to configure your ssl credentials and the location of your PuppetDB.
 
 By default the tool will use `$HOME/.puppetlabs/client-tools/puppetdb.conf` as
-it's configuration file if it exists.
+it's configuration file if it exists. You can also configure a global
+configuration (for all users) in `/etc/puppetlabs/client-tools/puppetdb.conf` to
```
Comment (senior) [inline README.md]: Do they know what this will look like for windows?

File: src/config.rs [hunk 4]
```diff
@@ -35,6 +42,15 @@ pub struct Config {
     pub token: Option<String>,
 }
 
+pub fn merge_configs(first: PdbConfigSection, second: PdbConfigSection) -> PdbConfigSection {
```
Comment (senior) [inline src/config.rs]: I think there should be some tests of this merge and the overlay from the global config section

File: src/config.rs [hunk 5]
```diff
@@ -15,6 +15,13 @@ pub fn default_config_path(mut home_dir: PathBuf) -> String {
     home_dir.to_str().unwrap().to_owned()
 }
 
+pub fn global_config_path() -> String {
+    let mut path = PathBuf::from("/etc/puppetlabs/client-tools");
+    path.push("puppetdb");
+    path.set_extension("conf");
+    path.to_str().unwrap().to_owned()
```
Comment (senior) [inline src/config.rs]: So it looks like what's going on here is path.to_str() is creating an &str of the path. Then the to_owned() call is copying the string reference to a String and giving ownership to the caller?

File: README.md (hunk 3 above)
Comment (ajroetker) [inline README.md]: I've not heard anything. I'll ping the puppet-acess folks to see what they know.

File: src/config.rs (hunk 5 above)
Comment (ajroetker) [inline src/config.rs]: Talked about this in standup but yes this is correct!

File: src/config.rs (hunk 4 above)
Comment (ajroetker) [inline src/config.rs]: I'll make a ticket for doing this in acceptance tests, but I want to wait on getting the FOSS and PE module PRs for configuring that stuff out of the box by default before adding those tests.

File: README.md (hunk 3 above)
Comment (briancain) [inline README.md]: @senior @ajroetker FYI: https://github.com/puppetlabs/puppet-specifications/blob/master/file_paths.md#puppet-agent-windows

File: src/config.rs [hunk 6]
```diff
@@ -15,6 +15,22 @@ pub fn default_config_path(mut home_dir: PathBuf) -> String {
...
+#[cfg(not(windows))]
+pub fn global_config_path() -> String {
+    let mut path = PathBuf::from("/etc/puppetlabs/client-tools");
+    path.push("puppetdb");
+    path.set_extension("conf");
+    path.to_str().unwrap().to_owned()
+}
+#[cfg(windows)]
+pub fn global_config_path() -> String {
+    let mut path = PathBuf::from("C:\\ProgramData\\PuppetLabs\\client-tools");
```
Comment (ajroetker) [inline src/config.rs]: @briancain @senior Added a conditional compiler flag for using the windows path here. Just fyi.

Comment (briancain) [inline src/config.rs]: Just a heads up, it looks like this folder location could be different depending on what version of windows the user is running this on. We were using a function from boost in c++ to determine the PROGRAMDATA folder location (which ended up having a bug) but you might want to look into that: https://tickets.puppetlabs.com/browse/PE-16040

File: src/lib.rs [hunk 7]
```diff
@@ -15,3 +30,45 @@ pub mod net;
 pub mod config;
 pub mod client;
 pub mod admin;
+
+#[cfg(windows)]
+pub mod windows {
```
Comment (ajroetker) [inline src/lib.rs]: @senior @briancain Alright I think I have the appropriate code now lifted from https://github.com/rust-lang-nursery/rustup.rs/blob/master/src/rustup-utils/src/raw.rs#L821and I got the GUID from the Microsoft website.

**Pull request merged**